node/src/components/linear_chain_sync.rs

```diff
@@ -147,53 +303,92 @@ where
+                    // the case then we have a bug.
+                    effect_builder
+                        .immediately()
+                        .event(move |_| Event::StartDownloadingDeploys)
+                }
+                BlockByHeightResult::FromPeer(block, peer) => {
+                    self.state.block_downloaded(&block);
+                    if block.height() != block_height {
+                        warn!(
+                            "Block height mismatch. Expected {} got {} from {}.",
+                            block_height,
+                            block.height(),
+                            peer
+                        );
+                        // NOTE: Signal misbehaving validator to networking layer.
+                        self.ban_peer(peer);
+                        return self.handle_event(
+                            effect_builder,
+                            rng,
+                            Event::GetBlockHeightResult(block_height, BlockByHeightResult::Absent),
+                        );
+                    }
+                    trace!(%block_height, "Downloaded linear chain block.");
+                    let next_height = block.height() + 1;
+                    self.add_block(*block);
+                    self.reset_peers(rng);
+                    let peer = self.random_peer_unsafe();
+                    fetch_block_at_height(effect_builder, peer, next_height)
+                }
+            },
+            Event::GetBlockHashResult(block_hash, fetch_result) => match fetch_result {
+                None => match self.random_peer() {
                     None => {
                         error!(%block_hash, "Could not download linear block from any of the peers.");
                         panic!("Failed to download linear chain.")
                     }
-                    Some(peer) => fetch_block(effect_builder, peer, block_hash),
+                    Some(peer) => fetch_block_by_hash(effect_builder, peer, block_hash),
                 },
                 Some(FetchResult::FromStorage(block)) => {
-                    // remember the era of the init block
-                    if Some(*block.hash()) == self.init_hash {
-                        self.init_block_era = Some(block.era_id());
-                    }
+                    self.state.block_downloaded(&block);
                     // We should be checking the local storage for linear blocks before we start
                     // syncing.
                     trace!(%block_hash, "Linear block found in the local storage.");
                     // If we found the linear block in the storage it means we should have all of
                     // its parents as well. If that's not the case then we have a bug.
                     effect_builder
                         .immediately()
-                        .event(move |_| Event::LinearChainBlocksDownloaded)
+                        .event(move |_| Event::StartDownloadingDeploys)
```

Comment:
Just forget about the local storage for now. This story is not implemented yet at all. It will be worked on in this sprint once storage does support returning max height block.

Reply:
Please add a `TODO` here then.
(I still don't understand why `StartDownloadingDeploys` will be correct here even once that is implemented, but we can discuss that in another PR, of course.)

Reply:
it might not be correct anymore but I don't want to think about it now, I don't know what the design then will look like.

Reply:
Resolving for now.